diffrax/integrate.py

```diff
@@ -195,6 +203,168 @@ def _apply_history(x):
+                        _tmid = _ta + 0.5 * (_tb - _ta)
+                        _gas = jnp.stack([jnp.sign(g(_ta)) for g in _gs])
+                        _gmids = jnp.stack([jnp.sign(g(_tmid)) for g in _gs])
+                        _gbs = jnp.stack([jnp.sign(g(_tb)) for g in _gs])
+                        _any_left = jnp.any(_gas != _gmids)
+                        _next_ta = jnp.where(_any_left, _ta, _tmid)
+                        _next_tb = jnp.where(_any_left, _tmid, _tb)
+                        _pred = (
+                            jnp.any(jnp.sum(_gas != _gbs, axis=1) > 1) | _step
+                            > max_step
+                        )
+                        return _next_ta, _next_tb, _pred, _step, max_step
+
+                    _init_val = (sub_tprev, sub_tnext, True, 0, 400)
+                    _final_val = lax.while_loop(_cond_fun, _body_fun, _init_val)
+                    _ta, _tb, _, _, _ = _final_val
+
+                    # Then do a more expensive Newton search
+                    # to find the first discontinuity.
+                    _discont_solver = NewtonNonlinearSolver(
+                        rtol=delays.rtol, atol=delays.atol
+                    )
+                    _disconts = []
+                    for g, delay in zip(_gs, flat_delays):
+                        changed_sign = jnp.sign(g(_ta)) != jnp.sign(g(_tb))
+                        if changed_sign.all() is False:
+                            pass
+                        else:
+                            _i = jnp.argmax(changed_sign)
+                            _d = state.discontinuities[_i]
+                            _h = (
+                                lambda t, args, delay=delay, _d=_d: t
+                                - delay(t, dense_discont.evaluate(t), args)
+                                - _d
+                            )
+                            _discont = _discont_solver(_h, _tb, args).root
+                            _disconts.append(_discont)
+                    if _discont.size == 0:
+                        return jnp.inf
+                    else:
+                        _disconts = jnp.stack(_disconts)
+                        valid_discont = (_disconts > sub_tprev) & (
+                            _disconts < sub_tnext
+                        )
+                        proxy_candidate = (
+                            _disconts * valid_discont
+                            + jnp.inf * jnp.invert(valid_discont) * _discont
+                        )
+                        best_candidate = jnp.min(proxy_candidate)
+                        best_candidate = jnp.where(
+                            jnp.isinf(best_candidate), sub_tnext, best_candidate
+                        )
+                        discont_update = jnp.where(
+                            jnp.isinf(best_candidate) | (best_candidate == sub_tnext),
+                            False,
+                            True,
+                        )
```

Comment:
Similar to commentary earlier, I think these pieces can be simplified.
(`where` instead of `* valid_discont`; and I think `jnp.isinf(best_candidate)` implies `jnp.invert(valid_discont)`?

Reply:
I agree, i'll try to find another way, i found this wacky procedure in order to get around the `ConcretizationError` in JAX.  Getting the correct breaking points is "fairly tricky" with all of the `jnp.inf`.
`best_candidate` here would be a float compared to `valid_discont` that tells us for each element of `tnext_candidate` which one is valid.